### ISSUE TITLE
Last Hidden Layers Embedding for Emotion Model

### ISSUE THREAD

**Issue by VSK02101996** [s3prl/s3prl#388]
How one can get last hidden layer representation when you pass the audio file as input ? Is there any direct way one can get that embeddings ?

**Comment by leo19941227** [comment]
Hello, I guess your question is about how to extract speech representation given an audio file?
If so, please refer to [our doc](https://s3prl.github.io/s3prl/tutorial/upstream_collection.html)

**Comment by VSK02101996** [comment]
import torch
from s3prl.nn import S3PRLUpstream

model = S3PRLUpstream("hubert")
model.eval()

with torch.no_grad():
    wavs = torch.randn(2, 16000 * 2)
    wavs_len = torch.LongTensor([16000 * 1, 16000 * 2])
    all_hs, all_hs_len = model(wavs, wavs_len)

for hs, hs_len in zip(all_hs, all_hs_len):
    assert isinstance(hs, torch.FloatTensor)
    assert isinstance(hs_len, torch.LongTensor)

    batch_size, max_seq_len, hidden_size = hs.shape
    assert hs_len.dim() == 1

This is the code to get representation ? right

**Comment by leo19941227** [comment]
Yes, the `all_hs` is the representation you want.

**Comment by VSK02101996** [comment]
Also the tensors we are getting are of all same size or different here ? I am actually building emotion classifier using https://github.com/KrishnaDN/Emotion-Recognition-Using-MSCNN-SPU ; so I want representation which I can directly append to mfcc features.

**Comment by leo19941227** [comment]
So far, all the tensor have the same shape. However, this is not guaranteed since the API the designed to be able to cope with future advanced models with different shapes in internal layers. Hence, you can know each layers' property according to [here](https://s3prl.github.io/s3prl/_autosummary/s3prl.nn.upstream.html#s3prl.nn.upstream.S3PRLUpstream). You can get the information for each layer, including hidden_size and stride (downsample rate from 16000 Hz)

**Comment by VSK02101996** [comment]
Also how your model performs for cross corpus setting ?

**Comment by leo19941227** [comment]
I believe this is also a research topic for SSL models transferability to unseen language / corpus... etc.
So far, it is shown in [SUPERB](https://arxiv.org/abs/2105.01051) they have quite good transferability on many different corpora & tasks including IEMOCAP (audio-only). However, I believe there are still other corpora they can not deal with, especially for the noisy ones, since they are usually pretrained on LibriSpeech, LibriLight with relatively clean speech.

**Comment by VSK02101996** [comment]
https://huggingface.co/superb/wav2vec2-base-superb-er Where I can get last layer hidden representation for this model ?

**Comment by VSK02101996** [comment]
Hi @leo19941227  which model have you used to train after getting wave2vec embeddings , also can you share the code for same ?

**Comment by leo19941227** [comment]
Specified by:

https://github.com/s3prl/s3prl/blob/087b27bfb5d3eb265a6d9a17bc843a55416d98ab/s3prl/downstream/emotion/config.yaml#L38
